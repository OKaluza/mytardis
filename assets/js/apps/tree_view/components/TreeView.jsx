import React, { useState, useEffect, Fragment } from 'react';

import { Treebeard, decorators } from 'react-treebeard';

import PropTypes from 'prop-types';
import { saveAs } from 'file-saver';
import Cookies from 'js-cookie';
import styles from './custom-theme';
import Header from './Header';
import Container from './Container';
import * as filters from './Filter';
import 'regenerator-runtime/runtime';
import { TreeDownloadButton } from './Download';


const TreeView = ({ datasetId, modified }) => {
  const [cursor, setCursor] = useState(false);
  const [data, setData] = useState([]);
  const [selectedCount, setSelectedCount] = useState(0);
  const fetchBaseDirs = () => {
    fetch(`/api/v1/dataset/${datasetId}/root-dir-nodes/`, {
      method: 'get',
      headers: {
        'Accept': 'application/json', // eslint-disable-line quote-props
        'Content-Type': 'application/json',
      },
    }).then(responseJson => (responseJson.json()))
      .then((response) => { setData(response); });
  };
  const fetchChildDirs = (node, dirPath) => {
    const encodedDir = encodeURIComponent(dirPath);
    fetch(`/api/v1/dataset/${datasetId}/child-dir-nodes/?dir_path=${encodedDir}`, {
      method: 'get',
      headers: {
        'Accept': 'application/json', // eslint-disable-line quote-props
        'Content-Type': 'application/json',
      },
    }).then(response => (response.json()))
      .then((childNodes) => {
        node.children = childNodes;
        node.toggled = true;
        if (node.selected) {
          node.children.forEach((childNode) => {
            childNode.selected = true;
          });
        }
        setData(Object.assign([], data));
      });
    //
  };
  useEffect(() => {
    fetchBaseDirs('');
  }, [datasetId, modified]);
  const onToggle = (node, toggled) => {
    // fetch children:
    // console.log(`on toggle ${node.name}`);
    if (toggled && node.children && node.children.length === 0) {
      fetchChildDirs(node, node.path);
      return;
    }
    if (cursor) {
      cursor.active = false;
    }
    node.active = true;
    if (node.children) {
      node.toggled = toggled;
    }
    setCursor(node);
    setData(Object.assign([], data));
  };
  const onFilterMouseUp = ({ target: { value } }) => {
    const filter = value.trim();
    if (!filter) {
      // set initial tree state:
      fetchBaseDirs('');
    }
    const filteredData = [];
    data.forEach((item) => {
      let filtered = filters.filterTree(item, filter);
      filtered = filters.expandFilteredNodes(filtered, filter);
      filteredData.push(filtered);
    });

    setData(filteredData);
  };
  const onSelect = (node) => {
    node.toggled = !node.toggled;
    if (node.selected) {
      // select all child nodes
      node.selected = false;
      setSelectedCount(selectedCount - 1);
      node.children.forEach((childNode) => {
        childNode.selected = false;
        setSelectedCount(selectedCount - 1);
      });
    } else {
      node.selected = true;
      setSelectedCount(selectedCount + 1);
    }
  };
  const downloadSelected = (event) => {
    let fileName = '';
    event.preventDefault();
    const formData = new FormData(event.target);
    console.log(data);
    let selectedData = [];
    data.forEach((item) => {
      const selected = filters.findSelected(item, []);
      selectedData = [...selectedData, ...selected];
    });
    console.log(selectedData);
    selectedData.forEach((id) => {
      if (typeof id === 'number') {
        formData.append('datafile', id.toString());
      }
    });
    fetch('/download/datafiles/', {
      method: 'POST',
      body: formData,
      headers: {
        'X-CSRFToken': Cookies.get('csrftoken'),
      },
    }).then((resp) => {
      console.log(resp);
      const disposition = resp.headers.get('Content-Disposition');
      console.log(disposition);
      if (disposition && disposition.indexOf('attachment') !== -1) {
        const filenameRegex = /filename[^;=\n]*=((['"]).*?\2|[^;\n]*)/;
        const matches = filenameRegex.exec(disposition);
        console.log(matches[1]);
        if (matches != null && matches[1]) {
          fileName = matches[1].replace(/['"]/g, '');
        }
      }
      resp.blob().then((fileContent) => {
        saveAs(fileContent, fileName);
      });
    });
  };
  return (
    <Fragment>
      <TreeDownloadButton
        count={selectedCount}
        onClick={downloadSelected}
      />
      <div style={styles}>
        <div className="input-group">
          <span className="input-group-text">
            <i className="fa fa-search" />
          </span>
          <input
            name="search-input"
            className="form-control"
            onKeyUp={onFilterMouseUp}
            placeholder="Search the tree..."
            type="text"
          />
        </div>
      </div>
      <div className="p-1" style={styles}>
        <Treebeard
          data={data}
          style={styles}
          onToggle={onToggle}
          onSelect={onSelect}
          decorators={{ ...decorators, Header, Container }}
          animation={false}
        />
      </div>
    </Fragment>
  );
};

TreeView.propTypes = {
  datasetId: PropTypes.string.isRequired,
  modified: PropTypes.string,
};

TreeView.defaultProps = {
  modified: '',
};

export default TreeView;
