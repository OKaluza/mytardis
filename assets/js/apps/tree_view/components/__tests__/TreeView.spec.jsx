import React from 'react';
import ReactDOM from 'react-dom';
import { act } from '@testing-library/react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TreeView from '../TreeView';


global.fetch = require('jest-fetch-mock');

configure({ adapter: new Adapter() });

const fakeTreeData = [{
  name: 'parent_1',
  path: 'parent_1',
  children: [],
},
{
  name: 'parent_2',
  path: 'parent_2',
  children: [],
},
{
  name: 'Parent.txt',
  id: 11985776,
},
{
  name: 'STORM-6.jpg',
  id: 11985840,
}];
const fakeChildData = [
  {
    name: 'child_1',
    path: 'parent_1/child_1',
    children: [],
  },
  {
    name: 'child_2',
    path: 'parent_1/child_2',
    children: [],
  },
  {
    name: 'Child_1.txt',
    id: 11985763,
  },
  {
    name: 'Child_2.txt',
    id: 11985764,
  }];

describe('renders initial tree view on page load', () => {
  let container = null;
  beforeEach(() => {
    jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve(fakeTreeData),
    }));
    container = document.createElement('div');
    document.body.appendChild(container);
  });
  afterEach(() => {
    // cleanup on exiting
    container.remove();
    container = null;
  });
  it('should render folder tree correctly', async () => {
    await act(async () => {
      ReactDOM.render(<TreeView datasetId="1234" modified="" />, container);
    });
    expect(container).toMatchSnapshot();
  });
  it('should render tree view with four child nodes', async () => {
    await act(async () => {
      ReactDOM.render(<TreeView datasetId="1234" modified="" />, container);
    });
    expect(container.querySelector('ul').children.length).toEqual(4);
    expect(container.querySelector('ul').children[0].textContent).toEqual('parent_1');
    expect(container.querySelector('ul').children[1].textContent).toEqual('parent_2');
    expect(container.querySelector('ul').children[2].textContent).toEqual('Parent.txt');
    expect(container.querySelector('ul').children[3].textContent).toEqual('STORM-6.jpg');
  });
});

describe('test rendering child nodes and filter', () => {
  let component = null;
  beforeEach(async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve(fakeTreeData),
    }));
    await act(async () => {
      component = mount(<TreeView datasetId="1234" modified="" />);
    });
    component.update();
  });
  it('should render child nodes when clicked on parent', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve(fakeChildData),
    }));
    await act(async () => {
      component.find('NodeHeader').first().simulate('click');
    });
    component.update();
    expect(component.find('NodeHeader')).toHaveLength(8);
    expect(component.find('Header').get(0).props.iconClass).toEqual('folder-open');
    expect(component.find('Header').get(0).props.node.children.length).toEqual(4);
    expect(component.find('Header').get(0).props.node.children[0].name).toEqual('child_1');
  });
  it('should filter tree node on search text change', async () => {
    await act(async () => {
      const searchField = component.find('input.form-control');
      searchField.simulate('keyUp', { target: { name: 'search-input', value: '2' } });
      // searchField.simulate('keyUp', { keyCode: 50 });
      setImmediate(() => {
        component.update();
        expect(component.find('NodeHeader')).toHaveLength(6);
      });
    });
  });
});
