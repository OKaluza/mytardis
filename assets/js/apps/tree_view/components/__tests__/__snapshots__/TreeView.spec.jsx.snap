// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders initial tree view on page load should render folder tree correctly 1`] = `
<div>
  <button
    class="btn btn-outline-secondary"
    count="0"
    style="margin-bottom: 12px; font-weight: bold;"
    title="Download"
    type="submit"
  >
    <i
      class="fa fa-download fa-sm mr-1 "
    />
    Select All
  </button>
  <div>
    <div
      class="input-group"
    >
      <span
        class="input-group-text"
      >
        <i
          class="fa fa-search"
        />
      </span>
      <input
        class="form-control"
        name="search-input"
        placeholder="Search the tree..."
        type="text"
      />
    </div>
  </div>
  <div
    class="p-1"
  >
    <ul
      class="css-1tl97mg"
    >
      <li
        class="css-ya4lz1"
      >
        <div>
          <div
            style="display: inline-block; vertical-align: top; color: rgb(0, 0, 0); padding-bottom: 2px;"
          >
            <div
              style="line-height: 20px; vertical-align: middle; font-size: 16px; cursor: pointer;"
            >
              <input
                class="datafile_checkbox"
                style="margin-right: 5px;"
                type="checkbox"
                value="parent_1"
              />
              <i
                class="fa fa-folder"
                style="margin-right: 5px; opacity: 0.6;"
              />
              parent_1
              
            </div>
          </div>
        </div>
        <div />
      </li>
      <li
        class="css-ya4lz1"
      >
        <div>
          <div
            style="display: inline-block; vertical-align: top; color: rgb(0, 0, 0); padding-bottom: 2px;"
          >
            <div
              style="line-height: 20px; vertical-align: middle; font-size: 16px; cursor: pointer;"
            >
              <input
                class="datafile_checkbox"
                style="margin-right: 5px;"
                type="checkbox"
                value="parent_2"
              />
              <i
                class="fa fa-folder"
                style="margin-right: 5px; opacity: 0.6;"
              />
              parent_2
              
            </div>
          </div>
        </div>
        <div />
      </li>
      <li
        class="css-ya4lz1"
      >
        <div>
          <div
            style="display: inline-block; vertical-align: top; color: rgb(0, 0, 0); padding-bottom: 2px;"
          >
            <div
              style="line-height: 20px; vertical-align: middle; font-size: 16px; cursor: pointer;"
            >
              <input
                class="datafile_checkbox"
                style="margin-right: 5px;"
                type="checkbox"
                value="Parent.txt"
              />
              <i
                class="fa fa-file-text"
                style="margin-right: 5px; opacity: 0.6;"
              />
              Parent.txt
              <a
                class="btn-sm"
                href="/api/v1/dataset_file/11985776/download/"
                style="color: Black;"
                title="Download"
              >
                <i
                  class="fa fa-download fa-sm"
                />
              </a>
            </div>
          </div>
        </div>
        <div />
      </li>
      <li
        class="css-ya4lz1"
      >
        <div>
          <div
            style="display: inline-block; vertical-align: top; color: rgb(0, 0, 0); padding-bottom: 2px;"
          >
            <div
              style="line-height: 20px; vertical-align: middle; font-size: 16px; cursor: pointer;"
            >
              <input
                class="datafile_checkbox"
                style="margin-right: 5px;"
                type="checkbox"
                value="STORM-6.jpg"
              />
              <i
                class="fa fa-file-text"
                style="margin-right: 5px; opacity: 0.6;"
              />
              STORM-6.jpg
              <a
                class="btn-sm"
                href="/api/v1/dataset_file/11985840/download/"
                style="color: Black;"
                title="Download"
              >
                <i
                  class="fa fa-download fa-sm"
                />
              </a>
            </div>
          </div>
        </div>
        <div />
      </li>
    </ul>
  </div>
</div>
`;
